src/listener/can_open_listener.py

```diff
@@ -64,26 +71,64 @@ def listen_to_network(self):
                 # And then get each variable's index and read it
                 sdo_index = sdo_object.od.index
                 try:
-                    # TODO: Only allows simple variables for now,
-                    #        needs to be able process complex variables.
-                    sdo_value = self.network[node_id].sdo.upload(
-                                sdo_index, 0)
-                    if(self._sdo_value_changed(sdo_index, sdo_value)):
-                        self.inform_interpreter(sdo_index)
+                    # Check for indices between 0x2000 and 0x5FFF
+                    # Because we except communication between those
+                    if (sdo_index >= 0x2000 and sdo_index <= 0x5FFF):
+                        if (type(sdo_object) == canopen.sdo.base.Array
+                           or type(sdo_object) == canopen.sdo.base.Record):
+                            self._read_complex_variable(
+                                self.network[node_id].sdo, sdo_index, node_id)
+                        elif (type(sdo_object) == canopen.sdo.base.Variable):
+                            self._read_simple_variable(
+                                self.network[node_id].sdo, sdo_index, node_id)
                 except SdoCommunicationError:
                     logging.error(f'The requested sdo ({hex(sdo_index)})'
-                                  ' is not received')
+                                  ' is not received!')
+
+    def _read_complex_variable(self, sdo_client, sdo_index, node_id):
+        for subindex in range(len(sdo_client[sdo_index]) + 1):
+            # Skips subindex 0 because there are no value changes around this
+            if(subindex != 0):
+                index_and_subindex = f'{sdo_index}sub{subindex}'
+                sdo_value = sdo_client.upload(sdo_index, subindex)
+                # Checks for every subindex if value changed
+                if(self._sdo_value_changed(index_and_subindex, sdo_value)):
+                    self.inform_interpreter(sdo_value,
+                                            sdo_client[sdo_index][subindex],
+                                            node_id)
+
+    def _read_simple_variable(self, sdo_client, sdo_index, node_id):
+        sdo_value = sdo_client.upload(sdo_index, 0)
+        if(self._sdo_value_changed(sdo_index, sdo_value)):
+            self.inform_interpreter(sdo_value, sdo_client[sdo_index].od.name,
+                                    node_id)
+
+    def _add_nodes(self, nodes):
+        for i in range(len(nodes)):
```

Comment:
volgens mij zit hier een foutje in. Op line 108 heb je `self.node_purposes.append(nodes[i]['node_purpose'])`. `['node_purpose']` is volgens mij letterlijk een string van "node_purpose" en niet een variabel die je inleest/de filepath.
Als het om een dictionary gaat, moet je volgens mij `dictName.get("key")` gebruiken, toch?

Zelfde geldt voor line 111 en 113 met `['eds_location']`

Reply:
Nee hoor, dit komt omdat het geen officiele python dictionary is maar dit is een object uit het yaml bestand. Ik heb de code getest en het werkt. Zie ook https://yaml.readthedocs.io/en/latest/example.html